### ISSUE TITLE
ROI returned by cvGetOptimalNewCameraMatrix has wront width & height

### ISSUE THREAD

**Issue by devernay** [opencv/opencv#24831]
### System Information

 OpenCV python version: 4.8.0.76
Operating System / Platform: macOS
Python version: 3.10.6

### Detailed description

I noticed [here](https://github.com/nerfstudio-project/nerfstudio/pull/2726#issuecomment-1881644844) that cvGetOptimalNewCameraMatrix returns (0,0,width-1,height-1) when the distortion parameters are zero.

Looking more closely at the code, I see that the internal function [icvGetRectangles](https://github.com/opencv/opencv/blob/9691a2dccf63f527122dc515b0a082d33e088572/modules/calib3d/src/calibration.cpp#L2518) finds the rectangles by sampling points in the (0,width-1)x(0,hieght-1) range, and then computes the output ROI sizes as [(maxX-minX,maxY-minY)](https://github.com/opencv/opencv/blob/9691a2dccf63f527122dc515b0a082d33e088572/modules/calib3d/src/calibration.cpp#L2556C41-L2556C57), which of course misses one pixel in each dimension!

I think the code in icvGetRectangles should add 1 pixel to both dimensions, because the max value should be included.

### Steps to reproduce

Test code:
```
import cv2
import numpy as np
# zero distortion
distortion_params = np.zeros(8)
width=1920
height=1080
ffactor=1.6
# intrinsics with a centered principal point
K1 = np.array([[width*ffactor, 0, (width-1)*0.5],
              [0,width*ffactor, (height-1)*0.5],
              [0,0,1]])
# intrinsics with a non-centered principal point
K2 = np.array([[width*ffactor, 0, (width-1)*0.5+0.05],
              [0,width*ffactor, (height-1)*0.5-0.04],
              [0,0,1]])
for K in [K1,K2]:
    newK, roi = cv2.getOptimalNewCameraMatrix(K, distortion_params, (width, height), 0)
    print(np.linalg.norm(K-newK))
    print(roi)
```
Output:
```
4.547473508864641e-13
(0, 0, 1919, 1079)
4.547473508864641e-13
(0, 0, 1919, 1079)
```
both rois should be (0, 0, 1920, 1080).

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by devernay** [comment]
note: previous icvGetRectangles fix, from 4.8.0: https://github.com/opencv/opencv/pull/23305

**Comment by f-dy** [comment]
bug introduced here: https://github.com/opencv/opencv/pull/23305/files#diff-2be6667fbe7cf15aaaa6c53808d9607f51b9fcf9ec85b116881eda0f883d45e5R2297

#23305 changed the pixel sampling from 0 to width-1, which is good, but then the ROI width should be maxX-minX+1, not maxX-minX

**Comment by vrabaud** [comment]
The error is actually a bit different. Adding `+1` is wrong as the rectangle is a `Rect_<double>` (switching it where you suggest gets the Calib3d_GetOptimalNewCameraMatrixNoDistortion.accuracy test to fail). The problem arises when converting the Rect to ints for the output. I am preparing a PR.

**Comment by devernay** [comment]
Yes it should do floor(x-0.5) on the minimum bound and ceil(x+0.5) on the second

**Comment by devernay** [comment]
Or something like that. To take into account the footprint of a pixel

**Comment by gchhablani** [comment]
Are there any updates on this one?
Would be nice to get this fix.
Thanks!

**Comment by Xuanrrr** [comment]
it seems haven't been fixed yet?

**Comment by asmorkalov** [comment]
yes.
